File: crates/engine-sim/python_ref/solver/state.py
"""Flat-array state representation for a single pipe.

Layout:
    q       : (n_cells + 2*n_ghost, 4) float64, conservative vars (ρA, ρuA, EA, ρYA)
    area    : (n_cells + 2*n_ghost,)  float64, cell-centred area [m^2]
    area_f  : (n_cells + 2*n_ghost + 1,) float64, face areas at interior+boundary faces
    dx      : float, uniform cell width [m]
    n_cells : int, number of real cells
    n_ghost : int, typically 2 for MUSCL-Hancock

Only real cells are evolved. Ghost cells are filled by BCs before each step.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

N_VARS = 4
I_RHO_A, I_MOM_A, I_E_A, I_Y_A = 0, 1, 2, 3  # indices into q axis 1


@dataclass
class PipeState:
    """Holds the flat arrays for one pipe domain.

    All arrays are float64 and owned by this object; numba kernels receive
    them by reference.

    Indexing:
        q[0]..q[n_ghost-1]               : left ghost cells
        q[n_ghost]..q[n_ghost+n_cells-1] : real cells
        q[n_ghost+n_cells]..             : right ghost cells
    """

    q: np.ndarray             # shape (N, 4)
    area: np.ndarray          # shape (N,)   cell-centred cross-section
    area_f: np.ndarray        # shape (N+1,) face cross-section
    dx: float
    n_cells: int
    n_ghost: int = 2
    gamma: float = 1.4         # frozen γ for the pipe. Mixture γ is handled
                               # via the cylinder/valve-BC side; for FSAE pipe
                               # flow this constant γ is adequate. Can be made
                               # variable later.
    R_gas: float = 287.0       # J/(kg·K), unburned air default
    wall_T: float = 320.0      # K
    hydraulic_D: np.ndarray = None  # shape (N,), hydraulic diameter [m]

    @property
    def n_total(self) -> int:
        return self.q.shape[0]

    def real_slice(self) -> slice:
        return slice(self.n_ghost, self.n_ghost + self.n_cells)


def make_pipe_state(
    n_cells: int,
    length: float,
    area_fn,              # callable A(x) [m^2] at cell centres and faces
    gamma: float = 1.4,
    R_gas: float = 287.0,
    wall_T: float = 320.0,
    n_ghost: int = 2,
    hydraulic_D_fn=None,  # callable D(x) [m]; default from area
) -> PipeState:
    """Build a PipeState for a pipe of given length and area profile.

    Cell centres at x_i = (i + 0.5) · dx, i = 0..n_cells-1.
    Face i+1/2 at x_{i+1/2} = i · dx, for i = 0..n_cells.
    """
    assert n_cells >= 4, "MUSCL-Hancock requires n_cells >= 4"
    dx = length / n_cells

    n_total = n_cells + 2 * n_ghost
    q = np.zeros((n_total, N_VARS), dtype=np.float64)
    area = np.zeros(n_total, dtype=np.float64)
    area_f = np.zeros(n_total + 1, dtype=np.float64)
    hyd_D = np.zeros(n_total, dtype=np.float64)

    # Real-cell centres + two ghost cells per side.
    for i in range(n_total):
        # Ghost cells to the left have i < n_ghost, to the right i >= n_ghost+n_cells
        i_real = i - n_ghost
        x_centre = (i_real + 0.5) * dx
        area[i] = float(area_fn(x_centre))
        if hydraulic_D_fn is None:
            # For a circular pipe, D = sqrt(4A/pi).
            hyd_D[i] = float(np.sqrt(max(4.0 * area[i] / np.pi, 0.0)))
        else:
            hyd_D[i] = float(hydraulic_D_fn(x_centre))

    for i in range(n_total + 1):
        i_real = i - n_ghost
        x_face = i_real * dx
        area_f[i] = float(area_fn(x_face))

    return PipeState(
        q=q, area=area, area_f=area_f, dx=dx,
        n_cells=n_cells, n_ghost=n_ghost,
        gamma=gamma, R_gas=R_gas, wall_T=wall_T,
        hydraulic_D=hyd_D,
    )


def set_uniform(state: PipeState, rho: float, u: float, p: float, Y: float = 0.0) -> None:
    """Initialise all cells (real + ghost) to a uniform primitive state."""
    gm1 = state.gamma - 1.0
    e_int = p / (gm1 * rho)   # specific internal energy
    E = rho * (e_int + 0.5 * u * u)  # total energy per unit volume
    A = state.area
    state.q[:, I_RHO_A] = rho * A
    state.q[:, I_MOM_A] = rho * u * A
    state.q[:, I_E_A]   = E * A
    state.q[:, I_Y_A]   = rho * Y * A


def set_left_right(
    state: PipeState, x0: float,
    rhoL: float, uL: float, pL: float, YL: float,
    rhoR: float, uR: float, pR: float, YR: float,
) -> None:
    """Initialise a left state for x < x0 and right state otherwise.

    Includes ghost cells; the caller's BC code will overwrite ghost states
    as needed.
    """
    gm1 = state.gamma - 1.0
    dx = state.dx
    n_ghost = state.n_ghost
    for i in range(state.n_total):
        i_real = i - n_ghost
        x_centre = (i_real + 0.5) * dx
        rho, u, p, Y = (rhoL, uL, pL, YL) if x_centre < x0 else (rhoR, uR, pR, YR)
        e_int = p / (gm1 * rho)
        E = rho * (e_int + 0.5 * u * u)
        A = state.area[i]
        state.q[i, I_RHO_A] = rho * A
        state.q[i, I_MOM_A] = rho * u * A
        state.q[i, I_E_A]   = E * A
        state.q[i, I_Y_A]   = rho * Y * A


def primitives_from_q_row(q_row: np.ndarray, A: float, gamma: float):
    """Convenience: get (ρ, u, p, Y) for a single cell from q and A."""
    rho = q_row[I_RHO_A] / A
    u = q_row[I_MOM_A] / (rho * A)
    E = q_row[I_E_A] / A
    kinetic = 0.5 * rho * u * u
    p = (gamma - 1.0) * (E - kinetic)
    Y = q_row[I_Y_A] / (rho * A)
    return rho, u, p, Y


def primitives_array(state: PipeState) -> np.ndarray:
    """Return (N,4) array of primitives (ρ, u, p, Y) for every cell."""
    n = state.n_total
    w = np.zeros((n, 4), dtype=np.float64)
    for i in range(n):
        rho, u, p, Y = primitives_from_q_row(state.q[i], state.area[i], state.gamma)
        w[i, 0] = rho
        w[i, 1] = u
        w[i, 2] = p
        w[i, 3] = Y
    return w


def total_mass(state: PipeState) -> float:
    """Integrate ρA dx over real cells only."""
    s = state.real_slice()
    return float(state.dx * state.q[s, I_RHO_A].sum())


def total_energy(state: PipeState) -> float:
    """Integrate EA dx over real cells only."""
    s = state.real_slice()
    return float(state.dx * state.q[s, I_E_A].sum())

File: crates/engine-sim/src/solver/state.rs
//! Flat-array pipe state representation.
//! Direct port of `solver/state.py`.
//!
//! Layout per cell (4 conservative variables):
//!     q[i*N_VARS + 0] = ρA
//!     q[i*N_VARS + 1] = ρuA
//!     q[i*N_VARS + 2] = EA
//!     q[i*N_VARS + 3] = ρYA

use std::f64::consts::PI;

pub const N_VARS: usize = 4;
pub const I_RHO_A: usize = 0;
pub const I_MOM_A: usize = 1;
pub const I_E_A:   usize = 2;
pub const I_Y_A:   usize = 3;

#[derive(Debug, Clone)]
pub struct PipeState {
    pub q: Vec<f64>,                // shape (n_total * N_VARS,)
    pub area: Vec<f64>,             // shape (n_total,)
    pub area_f: Vec<f64>,           // shape (n_total + 1,)
    pub hydraulic_d: Vec<f64>,      // shape (n_total,)
    pub dx: f64,
    pub n_cells: usize,
    pub n_ghost: usize,
    pub gamma: f64,
    pub r_gas: f64,
    pub wall_t: f64,
}

#[derive(Debug, Clone)]
pub struct ScratchBuffers {
    pub w:        Vec<f64>,   // (n_total * N_VARS)
    pub slopes:   Vec<f64>,
    pub w_pred_l: Vec<f64>,
    pub w_pred_r: Vec<f64>,
    pub flux:     Vec<f64>,   // ((n_total + 1) * N_VARS)
}

impl ScratchBuffers {
    pub fn for_pipe(state: &PipeState) -> Self {
        let n = state.n_total();
        Self {
            w:        vec![0.0; n * N_VARS],
            slopes:   vec![0.0; n * N_VARS],
            w_pred_l: vec![0.0; n * N_VARS],
            w_pred_r: vec![0.0; n * N_VARS],
            flux:     vec![0.0; (n + 1) * N_VARS],
        }
    }
}

impl PipeState {
    #[inline]
    pub fn n_total(&self) -> usize {
        self.n_cells + 2 * self.n_ghost
    }

    #[inline]
    pub fn real_range(&self) -> std::ops::Range<usize> {
        self.n_ghost..(self.n_ghost + self.n_cells)
    }

    /// q[i, k]
    #[inline]
    pub fn q_at(&self, i: usize, k: usize) -> f64 {
        self.q[i * N_VARS + k]
    }

    /// q[i, k] = v
    #[inline]
    pub fn q_set(&mut self, i: usize, k: usize, v: f64) {
        self.q[i * N_VARS + k] = v;
    }
}

/// Build a PipeState with cell-centred area from a callback. The callback
/// `area_fn(x)` returns A(x) in m² for any x ∈ [-n_ghost·dx, L + n_ghost·dx].
///
/// Ghost cells are real cells offset by n_ghost; their `area` and
/// `hydraulic_d` are evaluated at their (signed) x_centre, exactly as
/// the Python reference does.
pub fn make_pipe_state<F>(
    n_cells: usize,
    length: f64,
    mut area_fn: F,
    gamma: f64,
    r_gas: f64,
    wall_t: f64,
    n_ghost: usize,
) -> PipeState
where
    F: FnMut(f64) -> f64,
{
    assert!(n_cells >= 4, "MUSCL-Hancock requires n_cells >= 4");
    let dx = length / (n_cells as f64);
    let n_total = n_cells + 2 * n_ghost;
    let mut area = vec![0.0_f64; n_total];
    let mut area_f = vec![0.0_f64; n_total + 1];
    let mut hyd_d = vec![0.0_f64; n_total];

    for i in 0..n_total {
        let i_real = (i as i64) - (n_ghost as i64);
        let x_centre = ((i_real as f64) + 0.5) * dx;
        let a = area_fn(x_centre);
        area[i] = a;
        // Default hydraulic D: D = sqrt(4A/π)
        let arg = 4.0 * a / PI;
        let arg_safe = if arg > 0.0 { arg } else { 0.0 };
        hyd_d[i] = arg_safe.sqrt();
    }
    for i in 0..(n_total + 1) {
        let i_real = (i as i64) - (n_ghost as i64);
        let x_face = (i_real as f64) * dx;
        area_f[i] = area_fn(x_face);
    }

    PipeState {
        q: vec![0.0_f64; n_total * N_VARS],
        area,
        area_f,
        hydraulic_d: hyd_d,
        dx,
        n_cells,
        n_ghost,
        gamma,
        r_gas,
        wall_t,
    }
}

/// Initialise all cells (real + ghost) to a uniform primitive state.
pub fn set_uniform(state: &mut PipeState, rho: f64, u: f64, p: f64, y: f64) {
    let gm1 = state.gamma - 1.0;
    let e_int = p / (gm1 * rho);
    let big_e = rho * (e_int + 0.5 * u * u);
    let n_total = state.n_total();
    for i in 0..n_total {
        let a = state.area[i];
        state.q[i * N_VARS + I_RHO_A] = rho * a;
        state.q[i * N_VARS + I_MOM_A] = rho * u * a;
        state.q[i * N_VARS + I_E_A]   = big_e * a;
        state.q[i * N_VARS + I_Y_A]   = rho * y * a;
    }
}

#[allow(clippy::too_many_arguments)]
pub fn set_left_right(
    state: &mut PipeState, x0: f64,
    rho_l: f64, u_l: f64, p_l: f64, y_l: f64,
    rho_r: f64, u_r: f64, p_r: f64, y_r: f64,
) {
    let gm1 = state.gamma - 1.0;
    let dx = state.dx;
    let n_ghost = state.n_ghost;
    let n_total = state.n_total();
    for i in 0..n_total {
        let i_real = (i as i64) - (n_ghost as i64);
        let x_centre = ((i_real as f64) + 0.5) * dx;
        let (rho, u, p, y) = if x_centre < x0 {
            (rho_l, u_l, p_l, y_l)
        } else {
            (rho_r, u_r, p_r, y_r)
        };
        let e_int = p / (gm1 * rho);
        let big_e = rho * (e_int + 0.5 * u * u);
        let a = state.area[i];
        state.q[i * N_VARS + I_RHO_A] = rho * a;
        state.q[i * N_VARS + I_MOM_A] = rho * u * a;
        state.q[i * N_VARS + I_E_A]   = big_e * a;
        state.q[i * N_VARS + I_Y_A]   = rho * y * a;
    }
}

/// (ρ, u, p, Y) for a single cell.
#[inline]
pub fn primitives_from_cell(state: &PipeState, i: usize) -> (f64, f64, f64, f64) {
    let a = state.area[i];
    let rho = state.q[i * N_VARS + I_RHO_A] / a;
    let u = state.q[i * N_VARS + I_MOM_A] / (rho * a);
    let big_e = state.q[i * N_VARS + I_E_A] / a;
    let kinetic = 0.5 * rho * u * u;
    let p = (state.gamma - 1.0) * (big_e - kinetic);
    let y = state.q[i * N_VARS + I_Y_A] / (rho * a);
    (rho, u, p, y)
}

/// (N, 4) array of primitives (ρ, u, p, Y).
pub fn primitives_array(state: &PipeState) -> Vec<f64> {
    let n = state.n_total();
    let mut w = vec![0.0_f64; n * N_VARS];
    for i in 0..n {
        let (rho, u, p, y) = primitives_from_cell(state, i);
        w[i * N_VARS]     = rho;
        w[i * N_VARS + 1] = u;
        w[i * N_VARS + 2] = p;
        w[i * N_VARS + 3] = y;
    }
    w
}

/// Integrate ρA dx over real cells only.
pub fn total_mass(state: &PipeState) -> f64 {
    let mut s = 0.0_f64;
    for i in state.real_range() {
        s += state.q[i * N_VARS + I_RHO_A];
    }
    state.dx * s
}

pub fn total_energy(state: &PipeState) -> f64 {
    let mut s = 0.0_f64;
    for i in state.real_range() {
        s += state.q[i * N_VARS + I_E_A];
    }
    state.dx * s
}
